app/src/main/java/com/example/bottomnav/StructDecoder.java

```diff
@@ -1,47 +1,78 @@
+            Optional<DataDecoder> primitiveDecoder = PrimitiveDecoder.create(variable);
+            if (!primitiveDecoder.isPresent()) {
+                return Optional.empty();
+            }
+            decodedPrimatives.add(primitiveDecoder.get());
+        }
+        return Optional.of(new StructDecoder(decodedPrimatives));
     }
 
-    @Override
-    public String valueToString() {
-        StringBuilder message = new StringBuilder();
-        for (int i = 0; i < variables.size(); i++) {
-            String varName = variables.get(i).name;
-            message.append(decodedPrimatives.get(varName).valueToString());
-            message.append("\n");
-        }
-        return message.toString();
+    public StructDecoder(ArrayList<DataDecoder> decoded) {
+        decodedPrimitives = decoded;
     }
 
     @Override
-    public Object valueToRaw() {
-        return null;
+    public Optional<ArrayList<Optional>> decodeToRaw(byte[] payload) {
+        ArrayList<Optional> rawOptionals = new ArrayList<>();
+        for(DataDecoder decoder : decodedPrimitives) {
+            rawOptionals.add(decoder.decodeToRaw(payload));
+            payload = adjustPayload(payload, ((PrimitiveDecoder) decoder).typeSize);
+        }
+        return Optional.of(rawOptionals);
     }
 
+    
+    @RequiresApi(api = Build.VERSION_CODES.O)
     @Override
-    public String decode(Integer canId, byte[] payload) {
-        for (VariableContents variable : variables) {
-            Optional<DataDecoder> decoder = DataDecoder.createPrimitiveDecoder(variable);
-            if (decoder.isPresent()) {
-                decoder.get().decode(canId, payload);
-                decodedPrimatives.put(variable.name, decoder.get());
-                payload = adjustPayload(payload, ((PrimitiveDecoder) decoder.get()).getPacketSize());
+    public Optional<String> decodeToString(byte[] payload) {
+        ArrayList<String> decodedStrings = new ArrayList<>();
+        for(int index = 0; index < decodedPrimitives.size(); index++) {
+            PrimitiveDecoder decoder = (PrimitiveDecoder) decodedPrimitives.get(index);
+            byte[] dedicatedPayload = getDedicatedPayload(payload, decoder.typeSize);
+            if (index == decodedPrimitives.size() - 1) {
+                if (dedicatedPayload.length < payload.length) {
+                    decodedStrings.add("incorrectlyFormattedPayload: null");
+                    break;
+                }
             }
+            Optional<String> decoded = decoder.decodeToString(dedicatedPayload);
+            if (decoded.isPresent()) {
+                decodedStrings.add(decoded.get());
+            } else {
+                break;
+            }
+            payload = adjustPayload(payload, decoder.typeSize);
+        }
+        return Optional.of(String.join(",", decodedStrings));
+    }
+
+    // Splice function that return an array of bytes under specified
+    private byte[] getDedicatedPayload(byte[] payload, int typeSize){
```

Comment:
Would it make sense to have this return `Optional<byte[]>` instead of returning a truncated array?
This gives you stronger guarantees that if this returns a value, it will be exactly typeSize long

Reply:
Incorrect byte size is checked when a DecoderPrimitive decodes something.